### PR TITLE
mcp/examples: add PostgreSQL MCP server example

### PR DESCRIPTION
**Pull request by gk16**
Add standalone PostgreSQL MCP server example with development environment and tooling. This implementation provides read-only database access, schema introspection, and query capabilities while maintaining clean dependency separation from the main SDK.

## Motivation and Context

This is part of https://github.com/modelcontextprotocol/go-sdk/issues/33. This PR adds example of implementing mcp on [PostgreSQL](https://github.com/modelcontextprotocol/servers-archived/tree/main/src/postgres) from https://modelcontextprotocol.io/examples.
The PR provides a production-ready PostgreSQL MCP server example with proper deployment options, testing, and development tooling.

## How Has This Been Tested?

- 17 test cases, including unit tests, integration tests, and edge case handling
- Full docker-compose setup with PostgreSQL database and sample data
- Tested against live PostgreSQL with sample schema (users, products, orders tables)

## Breaking Changes

- **No Breaking Changes**: This is purely additive - a new example in a separate subproject

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation update

## Checklist

- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [x] I have added or updated documentation as needed

### Architecture Decisions
- **Independent Module**: Created `examples/postgres/go.mod` for this example
- **Read-Only Safety**: All database operations use read-only transactions to prevent accidental data modification

### Files Added
- `main.go` - PostgreSQL MCP server implementation with schema introspection and query tools
- `main_test.go` - Comprehensive test suite with mock database testing
- `go.mod/go.sum` - Independent module definition with local SDK reference
- `Dockerfile` - Multi-stage build for production deployment
- `docker-compose.yml` - Complete development environment with PostgreSQL
- `init.sql` - Sample database schema and data for testing
- `Makefile` - Build automation and development workflow
- `demo.sh` - Interactive demonstration script
- `README.md` - Complete documentation following Python server format
- `.env.example` - Configuration template
- `.gitignore` - Proper exclusions for Go and Docker artifacts

### Development Experience
- **Quick Start**: `make dev` sets up complete environment
- **Multiple Run Modes**: stdio (default) and HTTP (debugging) modes

### MCP Features Implemented
- **Resources**: Dynamic table discovery with schema information
- **Tools**: Read-only SQL query execution with transaction safety
- **Error Handling**: Proper MCP error responses and comprehensive error coverage
- **Resource URIs**: Standard format `postgres://host/table/schema`

This implementation serves as a reference for building production MCP servers while maintaining clean project architecture.

### CODE REVIEW

**Comment by jba** [comment]
Thanks for this large contribution. It's a lot of work, and we appreciate that. It's cool that you have a complete implementation, with a Makefile, shell scripts, Dockerfile, and so on.

This is definitely worth living somewhere. I'm just not sure it should live in this repo. There is a lot of code to maintain. What if we get a bug report against the Dockerfile or docker compose config? Or some bug about a runtime failure in a particular installation? I don't know if the maintainers of this repo are up to that sort of commitment.

We're going to think about this and get back to you.

/cc @findleyr @samthanawalla

**Comment by gk16** [comment]
> Thanks for this large contribution. It's a lot of work, and we appreciate that. It's cool that you have a complete implementation, with a Makefile, shell scripts, Dockerfile, and so on.
> 
> This is definitely worth living somewhere. I'm just not sure it should live in this repo. There is a lot of code to maintain. What if we get a bug report against the Dockerfile or docker compose config? Or some bug about a runtime failure in a particular installation? I don't know if the maintainers of this repo are up to that sort of commitment.
> 
> We're going to think about this and get back to you.
> 
> /cc @findleyr @samthanawalla

Thanks a lot for your nice words.
Yes, I agree. As more complete golang examples coming in, it would be nice to have a seperate repo to host them, like https://github.com/modelcontextprotocol/servers (they are mostly Python and TS examples)

File: examples/postgres/main.go [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+
+	pathComponents := []string{}
+	for _, component := range []string{resourceURL.Path} {
+		if component != "" {
+			pathComponents = append(pathComponents, component)
+		}
+	}
+
+	// Parse path: /tableName/schema
+	parts := []string{}
```
Comment (jba) [inline examples/postgres/main.go]: `var parts []string`

Here and everywhere else.

File: examples/postgres/main.go [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+
+	_ "github.com/lib/pq" // PostgreSQL driver
+	"github.com/modelcontextprotocol/go-sdk/mcp"
+)
+
+var httpAddr = flag.String("http", "", "if set, use streamable HTTP at this address, instead of stdin/stdout")
+
+const SCHEMA_PATH = "schema"
+
+type PostgresServer struct {
```
Comment (jba) [inline examples/postgres/main.go]: Move this and all related code to postgres_server.go.

File: examples/postgres/main.go [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+
+// readTableSchema handles reading table schema information
+func (ps *PostgresServer) readTableSchema(ctx context.Context, ss *mcp.ServerSession, params *mcp.ReadResourceParams) (*mcp.ReadResourceResult, error) {
+	resourceURL, err := url.Parse(params.URI)
+	if err != nil {
+		return nil, fmt.Errorf("invalid resource URI: %w", err)
+	}
+
+	pathComponents := []string{}
+	for _, component := range []string{resourceURL.Path} {
```
Comment (jba) [inline examples/postgres/main.go]: Why a loop here?

File: examples/postgres/main.go [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+	}
+
+	// Parse path: /tableName/schema
+	parts := []string{}
+	for _, part := range pathComponents {
+		if part == "/" {
+			continue
+		}
+		subParts := []string{}
+		for _, subPart := range []string{part} {
```
Comment (jba) [inline examples/postgres/main.go]: Again, why loops over one element?

File: examples/postgres/main.go [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+}
+
+// readTableSchema handles reading table schema information
+func (ps *PostgresServer) readTableSchema(ctx context.Context, ss *mcp.ServerSession, params *mcp.ReadResourceParams) (*mcp.ReadResourceResult, error) {
+	resourceURL, err := url.Parse(params.URI)
+	if err != nil {
+		return nil, fmt.Errorf("invalid resource URI: %w", err)
+	}
+
+	pathComponents := []string{}
```
Comment (jba) [inline examples/postgres/main.go]: It looks like there's a lot of parsing code here. It's not really clear what it's done. Split into a separate function, document that function clearly, and add tests.

File: examples/postgres/main.go [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+					subparts := []string{}
+					current := ""
+					for _, r := range p {
+						if r == '/' {
+							if current != "" {
+								subparts = append(subparts, current)
+								current = ""
+							}
+						} else {
+							current += string(r)
```
Comment (jba) [inline examples/postgres/main.go]: This can be quadratic. Consider using a []string and calling strings.Join at the end.

File: examples/postgres/main.go [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+}
+
+// QueryTool executes a read-only SQL query
+func (ps *PostgresServer) QueryTool(ctx context.Context, ss *mcp.ServerSession, params *mcp.CallToolParamsFor[QueryArgs]) (*mcp.CallToolResultFor[struct{}], error) {
+	sqlQuery := params.Arguments.SQL
+
+	// Start a read-only transaction
+	tx, err := ps.db.BeginTx(ctx, &sql.TxOptions{ReadOnly: true})
+	if err != nil {
+		return &mcp.CallToolResultFor[struct{}]{
```
Comment (jba) [inline examples/postgres/main.go]: You should be able to return an error from a ToolHandler and we'll turn it into a CallToolResult for that error.

File: examples/postgres/main.go [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+			Content: []mcp.Content{
+				&mcp.TextContent{Text: fmt.Sprintf("Failed to get columns: %v", err)},
+			},
+			IsError: true,
+		}, nil
+	}
+
+	// Prepare result slice
+	var results []map[string]interface{}
```
Comment (jba) [inline examples/postgres/main.go]: use `any`

File: examples/postgres/main_test.go [hunk 9]
```diff
@@ -0,0 +1,1021 @@
+// Copyright 2025 The Go MCP SDK Authors. All rights reserved.
+// Use of this source code is governed by an MIT-style
+// license that can be found in the LICENSE file.
+
+package main
```
Comment (jba) [inline examples/postgres/main_test.go]: Tests that are only about PostgresServer should move to postgres_server_test.go.

**Review by jba — CHANGES_REQUESTED** [review]

**Comment by jba** [comment]
The modelcontextprotocol/servers repo is not going to happen.

Here's what we'd like to do:

- Accept the Go code into this repo.
- You should host all the code, include scripts and docker config, in your own repo, and add a link here to it.

One concern is that the additional non-Go code is a maintenance burden for us. We are also concerned about security.

You can leave all the other code here for now if it helps you test my suggested changes.

Thanks again for this great piece of work!

File: examples/postgres/main.go (hunk 1 above)
Comment (gk16) [reply examples/postgres/main.go]: Got it, I have change all these declaration into this format.

File: examples/postgres/main.go (hunk 2 above)
Comment (gk16) [reply examples/postgres/main.go]: Done, `PostgresServer struct` and all its function are moved to `postgres_server.go`

File: examples/postgres/main.go (hunk 3 above)
Comment (gk16) [reply examples/postgres/main.go]: Fixed it.

File: examples/postgres/main.go (hunk 4 above)
Comment (gk16) [reply examples/postgres/main.go]: Fixed.

File: examples/postgres/main.go (hunk 5 above)
Comment (gk16) [reply examples/postgres/main.go]: Make sense, I have moved the parsing logic into a function called `parseResourceURI` and added unit tests.

File: examples/postgres/main.go (hunk 6 above)
Comment (gk16) [reply examples/postgres/main.go]: Thanks a lot. I did not know this problem before you pointed it out. I have optimized this part.

File: examples/postgres/main.go (hunk 7 above)
Comment (gk16) [reply examples/postgres/main.go]: I see, I have made them return a simple error `fmt.Errorf`

File: examples/postgres/main.go (hunk 8 above)
Comment (gk16) [reply examples/postgres/main.go]: Got it, fixed.

File: examples/postgres/main_test.go (hunk 9 above)
Comment (gk16) [reply examples/postgres/main_test.go]: Done.

**Comment by gk16** [comment]
> The modelcontextprotocol/servers repo is not going to happen.
> 
> Here's what we'd like to do:
> 
> * Accept the Go code into this repo.
> * You should host all the code, include scripts and docker config, in your own repo, and add a link here to it.
> 
> One concern is that the additional non-Go code is a maintenance burden for us. We are also concerned about security.
> 
> You can leave all the other code here for now if it helps you test my suggested changes.
> 
> Thanks again for this great piece of work!

Thank a lot for your review. I did learn a lot from it. 

I have addressed the comments. Please take another look when you have a chance.

As for the all other code, do you mean I can keep `main.go`, `postgres_server.go`, `postgres_server_test.go`, and `README.md` these 4 files here and move all other files into a seperate repo, right?

**Comment by jba** [comment]
> As for the all other code, do you mean I can keep main.go, postgres_server.go, postgres_server_test.go, and README.md these 4 files here and move all other files into a seperate repo, right?


Yes.

File: examples/postgres/postgres_server.go [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+
+	// Remove leading slash and split by '/'
+	trimmedPath := strings.TrimPrefix(urlPath, "/")
+
+	// Split path into components using strings.Split for better performance (O(n) vs O(n²))
+	var pathParts []string
+	if trimmedPath != "" {
+		pathParts = strings.Split(trimmedPath, "/")
+		// Filter out empty parts that might result from consecutive slashes
+		var filteredParts []string
```
Comment (jba) [inline examples/postgres/postgres_server.go:128]: Use slices.DeleteFunc.

File: examples/postgres/postgres_server_test.go [hunk 11]
```diff
@@ -39,6 +39,109 @@ func createMockPostgresServer(t *testing.T) (*PostgresServer, sqlmock.Sqlmock, f
...
+		{
+			name:          "invalid URI - empty path after slash",
+			resourceURI:   "postgres://localhost:5432/",
+			wantError:     true,
+			errorContains: "invalid resource URI format: expected /tableName/schema",
+		},
+		{
+			name:          "edge case - table name with special characters",
+			resourceURI:   "postgres://localhost:5432/testdb/test-table_123/schema",
+			expectedTable: "test-table_123",
```
Comment (jba) [inline examples/postgres/postgres_server_test.go:119]: Add a test with "///" to test removal of empty parts.

**Comment by jba** [comment]
Closing this per our new [timeout policy](https://github.com/modelcontextprotocol/go-sdk/blob/main/CONTRIBUTING.md#timeouts).
